plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/PulsarSchemaInfoProvider.java

```diff
@@ -0,0 +1,121 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+
+import static java.util.concurrent.CompletableFuture.completedFuture;
+
+/**
+ * Multi version schema info provider for Pulsar SQL leverage guava cache.
+ */
+public class PulsarSchemaInfoProvider
+        implements SchemaInfoProvider
+{
+    private static final Logger LOG = LoggerFactory.getLogger(PulsarSchemaInfoProvider.class);
+
+    private final TopicName topicName;
+
+    private final PulsarAdmin pulsarAdmin;
+
+    private final LoadingCache<BytesSchemaVersion, SchemaInfo> cache = CacheBuilder.newBuilder().maximumSize(100000)
+            .expireAfterAccess(30, TimeUnit.MINUTES).build(new CacheLoader<BytesSchemaVersion, SchemaInfo>()
+            {
+                @Override
+                public SchemaInfo load(BytesSchemaVersion schemaVersion) throws Exception
+                {
+                    return loadSchema(schemaVersion);
+                }
+            });
+
+    public PulsarSchemaInfoProvider(TopicName topicName,
+                                    PulsarAdmin pulsarAdmin)
+    {
+        this.topicName = topicName;
+        this.pulsarAdmin = pulsarAdmin;
+    }
+
+    @Override
+    public CompletableFuture<SchemaInfo> getSchemaByVersion(byte[] schemaVersion)
+    {
+        try {
+            if (null == schemaVersion) {
+                return completedFuture(cache.get(BytesSchemaVersion.of((new LongSchemaVersion(0)).bytes())));
+            }
+            return completedFuture(cache.get(BytesSchemaVersion.of(schemaVersion)));
+        }
+        catch (ExecutionException e) {
+            LOG.error("Can't get generic schema for topic {} schema version {}",
+                    topicName.toString(), new String(schemaVersion, StandardCharsets.UTF_8), e);
+            return FutureUtil.failedFuture(e.getCause());
+        }
+    }
+
+    @Override
+    public CompletableFuture<SchemaInfo> getLatestSchema()
+    {
+        try {
+            return completedFuture(pulsarAdmin.schemas()
+                    .getSchemaInfo(topicName.toString()));
+        }
+        catch (PulsarAdminException e) {
+            LOG.error("Can't get current schema for topic {}",
+                    topicName.toString(), e);
+            return FutureUtil.failedFuture(e.getCause());
+        }
+    }
+
+    @Override
+    public String getTopicName()
+    {
+        return topicName.getLocalName();
+    }
+
+    private SchemaInfo loadSchema(BytesSchemaVersion bytesSchemaVersion) throws PulsarAdminException
+    {
+        ClassLoader originalContextLoader = Thread.currentThread().getContextClassLoader();
+        try {
+//            Thread.currentThread().setContextClassLoader(InjectionManagerFactory.class.getClassLoader());
```

Comment:
remove

Reply:
actually it's needed, I thought it was causing dependency issue when start up the service so comment it out for testing, I've removed the comment.